keyvault/client/src/main/java/com/azure/keyvault/SecretClientBuilder.java

```diff
@@ -0,0 +1,205 @@
+ *
+ *    SecretClient secretClientWithCustomPipeline = SecretClient.builder()
+ *                                                  .pipeline(customHttpPipeline)
+ *                                                  .build()
+ * </pre>
+ *
+ * <p>The {@link HttpLogDetailLevel logdetailLevel}, multiple custom {@link HttpPipeline policies} and custom
+ * {@link HttpClient httpClient} be optionally configured in the {@link SecretClientBuilder}.</p>
+ *
+ * <pre>
+ *    SecretClient secretClient = SecretClient.builder()
+ *                                .vaultEndpoint("https://myvault.vault.azure.net/")
+ *                                .credentials(keyVaultCredentials)
+ *                                .httpLogDetailLevel(HttpLogDetailLevel.BODY_AND_HEADERS)
+ *                                .addPolicy(customPolicyOne)
+ *                                .addPolicy(customPolicyTwo)
+ *                                .httpClient(client)
+ *                                .build()
+ * </pre>
+ */
+public final class SecretClientBuilder {
+    private final List<HttpPipelinePolicy> policies;
+    private ServiceClientCredentials credentials;
+    private HttpPipeline pipeline;
+    private URL vaultEndpoint;
+    private HttpClient httpClient;
+    private HttpLogDetailLevel httpLogDetailLevel;
+    private RetryPolicy retryPolicy;
+    private String userAgent;
+
+    SecretClientBuilder() {
+        userAgent = String.format("Azure-SDK-For-Java/%s (%s)", AzureKeyVaultConfiguration.SDK_NAME, AzureKeyVaultConfiguration.SDK_VERSION);
+        retryPolicy = new RetryPolicy();
+        httpLogDetailLevel = HttpLogDetailLevel.NONE;
+        policies = new ArrayList<>();
+    }
+
+    /**
+     * Creates a {@link SecretClient} based on options set in the builder.
+     * Every time {@code build()} is called, a new instance of {@link SecretClient} is created.
+     *
+     * <p>If {@link SecretClientBuilder#pipeline(HttpPipeline) pipeline} is set, then the {@code pipeline} and
+     * {@link SecretClientBuilder#vaultEndpoint(String) serviceEndpoint} are used to create the
+     * {@link SecretClientBuilder client}. All other builder settings are ignored.</p>
+     *
+     * @return A SecretClient with the options set from the builder.
+     * @throws IllegalStateException If {@link SecretClientBuilder#credentials(ServiceClientCredentials)} or
+     * {@link SecretClientBuilder#vaultEndpoint(String)} have not been set.
+     */
+    public SecretClient build() {
+
+        if (pipeline != null) {
+            return new SecretClient(vaultEndpoint, pipeline);
+        }
+
+        if (credentials == null) {
+            throw new IllegalStateException(KeyVaultErrorCodeStrings.getErrorString(KeyVaultErrorCodeStrings.CREDENTIALS_REQUIRED));
+        }
+
+        if (vaultEndpoint == null) {
+            throw new IllegalStateException(KeyVaultErrorCodeStrings.getErrorString(KeyVaultErrorCodeStrings.VAULT_END_POINT_REQUIRED));
+        }
```

Comment:
You should null check vaultEndpoint before pipeline, because you use vaultEndpoint in the pipeline conditional code above.

Reply:
correct, updated.